src/tests/test_core_python.py

```diff
@@ -1,70 +1,99 @@
+        for key in list(info.keys()):
+            if key.endswith(browser):
+                del info[key]
+    return info
 
-    if any(flag.startswith("segfault") for flag in error_flags):
-        pytest.skip('known segfault: "{}"'.format(" ".join(error_flags)))
 
-    if error_flags:
+def possibly_skip_test(request, info: dict[str, Any]) -> dict[str, Any]:
+    for reason in (
+        reason for [flag, reason] in info.items() if flag.startswith("segfault")
+    ):
+        pytest.skip(f"known segfault: {reason}")
+
+    for reason in (
+        reason for [flag, reason] in info.items() if flag.startswith("xfail")
+    ):
         if request.config.option.run_xfail:
             request.applymarker(
                 pytest.mark.xfail(
                     run=False,
-                    reason='known failure: "{}"'.format(" ".join(error_flags)),
+                    reason=f"known failure: {reason}",
                 )
             )
         else:
-            pytest.xfail('known failure: "{}"'.format(" ".join(error_flags)))
+            pytest.xfail(f"known failure: {reason}")
+    return info
+
+
+def test_cpython_core(main_test, selenium, request):
+    [name, info] = main_test
+    info = filter_info(info, selenium.browser)
+    possibly_skip_test(request, info)
+
+    ignore_tests = info.get("skip", [])
+    if not isinstance(ignore_tests, list):
+        raise Exception("Invalid python_tests.yaml entry: 'skip' should be a list")
 
     selenium.load_package(list(UNVENDORED_STDLIB_MODULES))
     try:
         selenium.run(
+            f"""
+import platform
+from test import libregrtest
+
+platform.platform(aliased=True)
+import _testcapi
+if hasattr(_testcapi, "raise_SIGINT_then_send_None"):
+    # This uses raise() which doesn't work.
+    del _testcapi.raise_SIGINT_then_send_None
+
+try:
+    libregrtest.main(["{name}"], ignore_tests={ignore_tests}, verbose=True, verbose3=True)
+except SystemExit as e:
+    if e.code != 0:
+        raise RuntimeError(f"Failed with code: {{e.code}}")
             """
-            from test.libregrtest import main
-            import unittest.case
-            try:
-                main(['{}'], verbose=True, verbose3=True)
-            except SystemExit as e:
-                if e.code != 0:
-                    raise RuntimeError(f'Failed with code: {{e.code}}')
-            except unittest.case.SkipTest:
-                pass
-            """.format(
-                name
-            )
         )
     except selenium.JavascriptException:
         print(selenium.logs)
         raise
 
 
+def get_test_info(test):
+    if isinstance(test, dict):
+        [name, info] = next(iter(test.items()))
+    else:
+        name = test
+        info = {}
+    return [name, info]
+
+
+@cache
+def get_tests():
+    with open(Path(__file__).parent / "python_tests.yaml") as file:
+        data = yaml.load(file, Loader)
+
+    return {name: info for test in data for [name, info] in [get_test_info(test)]}
```

Comment:
Maybe making this a generator that yields `name, info` would be a bit more readable and it should work with the rest of the code.